Build failure: nethack

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

`nix-build -A nethack`

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/318063997

### Relevant log output

```console
make[1]: Entering directory '/build/NetHack-3.6.7/src'
touch ../src/config.h-t
gcc -O -I../include -DNOTPARMDECL -DDLB -DCOMPRESS=\"/nix/store/0hv4z5s3r1h4lmvn0427mlxjxgvg34nr-gzip-1.14/bin/gzip\" -DCOMPRESS_EXTENSION=\".gz\" -DSYSCF -DSYSCF_FILE=\"/nix/store/dwqj4ib5vmqr15lfm0n7i9nvqjnifa98-nethack-3.6.7/games/lib/nethackdir/sysconf\" -DSECURE -DTIMED_DELAY -DHACKDIR=\"/nix/store/dwqj4ib5vmqr15lfm0n7i9nvqjnifa98-nethack-3.6.7/games/lib/nethackdir\" -DDUMPLOG -DCONFIG_ERROR_SECURE=FALSE -DCURSES_GRAPHICS -c monst.c
In file included from ../include/unixconf.h:323,
                 from ../include/global.h:136,
                 from ../include/config.h:580,
                 from monst.c:6:
/nix/store/dj43clc5ff7jjnfmhbaj6q4q0h8kpfpm-glibc-2.40-66-dev/include/stdlib.h:601:13: error: conflicting types for 'srand48'; have 'void(long int)'
  601 | extern void srand48 (long int __seedval) __THROW;
      |             ^~~~~~~
In file included from ../include/unixconf.h:320:
../include/system.h:99:8: note: previous declaration of 'srand48' with type 'void(void)'
   99 | E void srand48();
      |        ^~~~~~~
In file included from ../include/unixconf.h:324:
/nix/store/dj43clc5ff7jjnfmhbaj6q4q0h8kpfpm-glibc-2.40-66-dev/include/unistd.h:464:21: error: conflicting types for 'sleep'; have 'unsigned int(unsigned int)'
  464 | extern unsigned int sleep (unsigned int __seconds);
      |                     ^~~~~
../include/system.h:358:12: note: previous declaration of 'sleep' with type 'unsigned int(void)'
  358 | E unsigned sleep();
      |            ^~~~~
make[1]: *** [Makefile:659: monst.o] Error 1
make[1]: Leaving directory '/build/NetHack-3.6.7/src'
make: *** [Makefile:161: nethack] Error 2
```

### Additional context

Possibly related https://github.com/NixOS/nixpkgs/issues/475479

### System metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.18.2, NixOS, 26.05 (Yarara), 26.05.20251228.c0b0e0f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.31.2+1`
 - nixpkgs: `/nix/store/72ym7zz89wa91k3g4vz2g97pma1n51zy-xzl5dhddgjxp2p18h5qpkhvlkp8128j1-source`
```

### Notify maintainers



---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

[^1]

[^1]: Please consider [giving up MS GitHub](https://sfconservancy.org/GiveUpGitHub/) or offering a non-proprietary, non-US-corporate-controlled mirror for this free software project. I wish to delete this Microsoft account in the future, but I need more projects like this to support alternative methods to send patches & contribute.

--- a/pkgs/by-name/ne/nethack/function-parameters.patch
+++ b/pkgs/by-name/ne/nethack/function-parameters.patch
@@ -1,0 +1,86 @@
+diff --git a/include/system.h b/include/system.h
+index a2cb51f..9f40fdd 100644
+--- a/include/system.h
++++ b/include/system.h
+@@ -60,17 +60,17 @@ typedef long off_t;
+ #endif
+ #ifndef SIG_RET_TYPE
+ #if defined(NHSTDC) || defined(POSIX_TYPES) || defined(OS2) || defined(__DECC)
+-#define SIG_RET_TYPE void (*)()
++#define SIG_RET_TYPE void (*)(int)
+ #endif
+ #endif
+ #ifndef SIG_RET_TYPE
+ #if defined(ULTRIX) || defined(SUNOS4) || defined(SVR3) || defined(SVR4)
+ /* SVR3 is defined automatically by some systems */
+-#define SIG_RET_TYPE void (*)()
++#define SIG_RET_TYPE void (*)(int)
+ #endif
+ #endif
+ #ifndef SIG_RET_TYPE /* BSD, SIII, SVR2 and earlier, Sun3.5 and earlier */
+-#define SIG_RET_TYPE int (*)()
++#define SIG_RET_TYPE int (*)(int)
+ #endif
+ 
+ #if !defined(__cplusplus) && !defined(__GO32__)
+@@ -96,7 +96,7 @@ E long NDECL(lrand48);
+ E void FDECL(srand48, (long));
+ #else
+ E long lrand48();
+-E void srand48();
++E void srand48(long);
+ #endif /* MACOSX */
+ #endif /* BSD || ULTRIX || RANDOM */
+ 
+@@ -352,10 +352,10 @@ E char *FDECL(memset, (char *, int, int));
+ #endif /* MICRO */
+ 
+ #if defined(BSD) && defined(ultrix) /* i.e., old versions of Ultrix */
+-E void sleep();
++E void sleep(unsigned);
+ #endif
+ #if defined(ULTRIX) || defined(SYSV)
+-E unsigned sleep();
++E unsigned sleep(unsigned);
+ #endif
+ #if defined(HPUX)
+ E unsigned int FDECL(sleep, (unsigned int));
+@@ -519,7 +519,7 @@ E char *FDECL(tgoto, (const char *, int, int));
+ #else
+ #if !(defined(HPUX) && defined(_POSIX_SOURCE))
+ E int FDECL(tgetent, (char *, const char *));
+-E void FDECL(tputs, (const char *, int, int (*)()));
++E void FDECL(tputs, (const char *, int, int (*)(int)));
+ #endif
+ E int FDECL(tgetnum, (const char *));
+ E int FDECL(tgetflag, (const char *));
+diff --git a/include/winX.h b/include/winX.h
+index a1a5605..1d6b84b 100644
+--- a/include/winX.h
++++ b/include/winX.h
+@@ -279,7 +279,7 @@ typedef struct {
+ } AppResources;
+ 
+ E AppResources appResources;
+-E void (*input_func)();
++E void (*input_func)(Widget, XEvent *, String *, Cardinal *);
+ 
+ extern struct window_procs X11_procs;
+ 
+diff --git a/include/xwindow.h b/include/xwindow.h
+index 8b9cfa0..8307c25 100644
+--- a/include/xwindow.h
++++ b/include/xwindow.h
+@@ -76,8 +76,10 @@
+ #define XtNexposeCallback "exposeCallback"
+ #define XtNresizeCallback "resizeCallback"
+ 
+-extern XFontStruct *WindowFontStruct(/* Widget */);
+-extern Font WindowFont(/* Widget */);
++struct Widget;
++
++extern XFontStruct *WindowFontStruct(Widget);
++extern Font WindowFont(Widget);
+ 
+ #define XtCWindowResource "WindowResource"
+ #define XtCRows "Rows"

--- a/pkgs/by-name/ne/nethack/package.nix
+++ b/pkgs/by-name/ne/nethack/package.nix
@@ -67,6 +67,11 @@ stdenvUsed.mkDerivation (finalAttrs: {
     hash = "sha256-mM9n323r+WaKYXRaqEwJvKs2Ll0z9blE7FFV1E0qrLI=";
   };
 
+  patches = [
+    # Newer GCC rejects function declarations without explicit parameters.
+    ./function-parameters.patch
+  ];
+
   buildInputs = [
     ncurses
   ]

